connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -16,56 +16,322 @@
+     * and the {@link TopicAdmin} used by that store.
+     *
+     * <p>The worker-global offset store is not modified as it may be used for other connectors that either already exist,
+     * or will be created, on this worker.
+     */
     @Override
     public void stop() {
-        // TODO
+        // Worker offset store should not be stopped as it may be used for multiple connectors
+        connectorStore.ifPresent(OffsetBackingStore::stop);
+        connectorStoreAdmin.ifPresent(TopicAdmin::close);
     }
 
+    /**
+     * Get the offset values for the specified keys.
+     *
+     * <p>If configured to use a connector-specific offset store, priority is given to the values contained in that store,
+     * and the values in the worker-global offset store (if one is provided) are used as a fallback for keys that are not
+     * present in the connector-specific store.
+     *
+     * <p>If not configured to use a connector-specific offset store, only the values contained in the worker-global
+     * offset store are returned.
+
+     * @param keys list of keys to look up
+     * @return future for the resulting map from key to value
+     */
     @Override
     public Future<Map<ByteBuffer, ByteBuffer>> get(Collection<ByteBuffer> keys) {
-        // TODO
-        return workerStore.get(keys);
+        Future<Map<ByteBuffer, ByteBuffer>> workerGetFuture = getFromStore(workerStore, keys);
+        Future<Map<ByteBuffer, ByteBuffer>> connectorGetFuture = getFromStore(connectorStore, keys);
+
+        return new Future<Map<ByteBuffer, ByteBuffer>>() {
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning) {
+                return workerGetFuture.cancel(mayInterruptIfRunning)
+                        || connectorGetFuture.cancel(mayInterruptIfRunning);
+            }
+
+            @Override
+            public boolean isCancelled() {
+                return workerGetFuture.isCancelled()
+                        || connectorGetFuture.isCancelled();
+            }
+
+            @Override
+            public boolean isDone() {
+                return workerGetFuture.isDone()
+                        && connectorGetFuture.isDone();
+            }
+
+            @Override
+            public Map<ByteBuffer, ByteBuffer> get() throws InterruptedException, ExecutionException {
+                Map<ByteBuffer, ByteBuffer> result = new HashMap<>(workerGetFuture.get());
+                result.putAll(connectorGetFuture.get());
+                return result;
+            }
+
+            @Override
+            public Map<ByteBuffer, ByteBuffer> get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+                long endTime = time.milliseconds() + unit.toMillis(timeout);
+                Map<ByteBuffer, ByteBuffer> result = new HashMap<>(workerGetFuture.get(timeout, unit));
+                timeout = Math.max(1, endTime - time.milliseconds());
```

Comment:
It's quite confusing that `timeout` is initially in `unit`, but in this reassignment becomes ms. I think maybe a new `timeoutRemainingMs` would be clearer.

Reply:
👍 done.